### ISSUE TITLE
How to recover node information from a trip?

### ISSUE THREAD

**Issue by mwoehlke-kitware** [valhalla/valhalla#2460]
How do I recover information about nodes from a `TripLeg`?

Ultimately, I need to determine the OSM node ids that make up a trip. I understand (#2172, #2083) that is probably not directly possible, but I can't figure out how to even approach the problem. In particular, it seems like I should at *least* be able to get back the intersections (Valhalla nodes) that the trip passes through, from which I'm hoping I can get lat/lon which I can use to perform a brute-force lookup on the OSM nodes. (I already have the OSM way ids, so a brute-force approach should only require checking a few, maybe a few dozen nodes, which should be feasible if inelegant.)

How can I extract something more than graph-connectivity and way id information from a `TripLeg`?

**Comment by dnesbitt61** [comment]
You can use trace_attributes to do this. See the map matching API document. This returns a list of edges along the matched path. Each edge will have a begin (and end) shape index which will allow you to index into the geometry/shape of the matched path to retrieve the node or intersection lat,lng.

**Comment by mwoehlke-kitware** [comment]
> Each edge will have a begin (and end) shape index

Ah, I'd missed those. Still...

> will allow you to index into the geometry/shape of the matched path to retrieve the node or intersection lat,lng

...how do I do *this*? The only node retrieval API I could find requires tile ids, and I don't know how to get those.

BTW, I am *not* using the high-level API (although if that's all you can point me to, I can probably reverse-engineer what the high-level API *does*; I've already been doing that). Reverse-decomposing from XML is a waste of CPU cycles. Right now I have a `valhalla::TripLeg` which I'm walking. (I see the shape indices now, but don't know what to do with them...)

**Comment by dnesbitt61** [comment]
Ah, you have the protocol buffer so this should be straightforward. Each TripLeg has an encoded shape. You can decode this and then use the edge within the Node (this is the edge leaving the node along the route path, intersecting edges are other edges that intersect the path at the node) to get the begin_shape_index which will allow you to get the lat,lng from the decoded shape vector.

**Comment by mwoehlke-kitware** [comment]
Okay, I was able to use `midgard::decode` to turn the shape string into points. From there, it seems my only option is to brute-force match the points against an OSM database :weary:. Oh, well, that's doable.

One thing I noticed is that Valhalla apparently drops some OSM nodes? For "overly detailed" ways, that shouldn't be an issue, but can I trust it will never drop a node with more than two edges and/or with more than one OSM way?
